clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +916,226 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
   OS << "\n} // Bultin2Qual\n";
 }
 
+JSONBuiltinInterfaceEmitter::TypeDesc::TypeDesc(const Record *T)
+    : Name(T->getValueAsString("Name")),
+      VecWidth(T->getValueAsInt("VecWidth")),
+      ElementSize(T->getValueAsInt("ElementSize")),
+      IsPointer(T->getValueAsBit("IsPointer")),
+      IsConst(T->getValueAsBit("IsConst")),
+      IsVolatile(T->getValueAsBit("IsVolatile")) {
+  if (T->getValueAsBit("IsInteger")) {
+    if (ElementSize == 1)
+      ElementType = BOOL;
+    else
+      ElementType = T->getValueAsBit("IsSigned") ? INT : UINT;
+  }
+  if (T->getValueAsBit("IsFloat"))
+    ElementType = FLOAT;
+  if (!T->isSubClassOf("FundamentalType") && !T->isSubClassOf("VectorType"))
+    ElementType = BUILTIN;
+
+  AddrSpace = StringSwitch<const char *>(T->getValueAsString("AddrSpace"))
+                  .Case("clang::LangAS::Default", "")
+                  .Case("clang::LangAS::opencl_private", "__private")
+                  .Case("clang::LangAS::opencl_global", "__global")
+                  .Case("clang::LangAS::opencl_constant", "__constant")
+                  .Case("clang::LangAS::opencl_local", "__local")
+                  .Default("__generic");
+}
+
+std::string JSONBuiltinInterfaceEmitter::TypeDesc::GetBaseTypeAsStr() const {
+  if (Name == "void")
+    return Name.str();
+
+  llvm::SmallString<32> Buffer;
+  {
+    llvm::raw_svector_ostream TypeStr(Buffer);
+    TypeStr << TypePrefix;
+    if (VecWidth != 1)
+      TypeStr << "_vec" << VecWidth;
+    switch (ElementType) {
+    case BOOL:
+      TypeStr << "_bool_t";
+      break;
+    case INT:
+      TypeStr << "_int" << ElementSize << "_t";
+      break;
+    case UINT:
+      TypeStr << "_uint" << ElementSize << "_t";
+      break;
+    case FLOAT:
+      TypeStr << "_fp" << ElementSize << "_t";
+      break;
+    case BUILTIN:
+      TypeStr << "_" << Name;
+      break;
+    }
+  }
+
+  return static_cast<std::string>(Buffer);
+}
+
+void JSONBuiltinInterfaceEmitter::Emit() {
+
+  // Parse the Records to populate the internal lists.
+  GetOverloads();
+  GroupBySignature();
+  ExpandTypes();
+
+  EmitBuiltins();
+}
+
+// Generate all possible types represented by the input record.
+void JSONBuiltinInterfaceEmitter::ExpandType(const Record *Ty) {
+  auto ExpansionIt = ExpandedTypes.find(Ty);
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVectorImpl<std::string> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    const Record *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (std::string &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += " ";
+        TypeStr += TypeDesc.AddrSpace;
```

Comment:
Since AddrSpace can be empty, do we want to omit one of the spaces?  Alternatively, perhaps AddrSpace can contain the leading or trailing space?

Reply:
Added the space to AddrSpace